pkg/volume/iscsi/iscsi_util.go

```diff
@@ -447,6 +547,22 @@ func (util *ISCSIUtil) DetachDisk(c iscsiDiskUnmounter, mntPath string) error {
 		// Logout may fail as no session may exist for the portal/IQN on the specified interface.
 		iface, found = extractIface(mntPath)
 	}
+
+	// Delete all the scsi devices and any multipath devices after unmounting
+	if err = deleteDevices(c); err != nil {
+		glog.Warningf("iscsi detach disk: failed to delete devices\nError: %v", err)
+		// Fall through -- even if deleting fails, a logout may fix problems
+	}
+	cnt--
+	if cnt != 0 {
+		return nil
+	}
+	// if device is no longer used, see if need to logout the target
+	refCount, err := getDevicePrefixRefCount(c.mounter, prefix)
```

Comment:
not sure if it is in scope of this PR, but IMHO kubelet shouldn't try to logout from iSCSI target. iSCSI session could be initiated at the host level before kubelet starts, logging out would yank iSCSI devices which are needed for host, but not needed by kubelet.

Reply:
This is a reasonable point, but the counter argument is that if kubelet never logs out of sessions, then we will leak iscsi sessions over time. Some storage controllers (Solidfire, for example) create a new session for each PV, so over the course of several weeks of running workloads, hundreds or thousands of defunct iSCSI sessions could accumulate.

It think it's better to document that, if you're using iSCSI for the root FS or some other host service that kubernetes doesn't know about, it must not share the iSCSI session with any storage backend. It would not be hard for an admin to set up multiple iSCSI targets to avoid that type of sharing.